dandi/files/zarr.py

```diff
@@ -153,25 +152,25 @@ def stat(self) -> ZarrStat:
 
         def dirstat(dirpath: LocalZarrEntry) -> ZarrStat:
             # Avoid heavy import by importing within function:
-            from dandi.support.digests import md5file_nocache
+            from dandi.support.digests import checksum_zarr_dir, md5file_nocache
 
             size = 0
-            dir_md5s = {}
-            file_md5s = {}
+            dir_info = {}
+            file_info = {}
             files = []
             for p in dirpath.iterdir():
                 if p.is_dir():
                     st = dirstat(p)
                     size += st.size
-                    dir_md5s[p.name] = (st.digest.value, st.size)
+                    dir_info[p.name] = (st.digest.value, st.size)
                     files.extend(st.files)
                 else:
                     size += p.size
-                    file_md5s[p.name] = (md5file_nocache(p.filepath), p.size)
+                    file_info[p.name] = (md5file_nocache(p.filepath), p.size)
                     files.append(p)
             return ZarrStat(
                 size=size,
-                digest=Digest.dandi_zarr(get_checksum(file_md5s, dir_md5s)),
+                digest=Digest.dandi_zarr(checksum_zarr_dir(file_info, dir_info)),
```

Comment:
do we really need to "manually" loop here via `dirpath.iterdir()` or could make use of https://github.com/dandi/zarr_checksum#python

```python
compute_zarr_checksum(yield_files_local("local_path"))
```
? i.e. could we reuse `yield_files_local` to collect `files` and then just pass to `compute_zarr_checksum` instead of reimplementing it as `checksum_zarr_dir`?

Reply:
hm, that sounds like a possible cause of differences/confusion which we ideally clarify/unify while at it.

I started writing items below, but then looked at the zarr-python implementation of traversal at https://github.com/zarr-developers/zarr-python/blob/HEAD/zarr/storage.py#L1174  and it seems to list all files really. Although "not guaranteed", I guess, but are you sure that it actually excludes any "extraneous" files @jwodder?

some thoughts I started to compose:

- in dandi-cli we typically (unless DEVEL option allow_any_path or bids datasets) upload only recognized by us files too, i.e. by default we ignore "extraneous" files, so `zarr_checksum`'s default (and only ATM) behavior is somewhat inline with that, while our own support of `zarr` diverges from it
- in principle we should not silently ignore extraneous files within `.zarr` -- but I guess zarr validation doesn't fail (does not care) about them right?
- do we have any knowledge of such extraneous files uploaded already? might be good to know and since we have dandizarrs datalad datasets on drogon -- we might be able to "quickly" (nothing is quick when dealing with that many files) check.

Reply:
It does seem that `NestedDirectoryStore.keys` essentially calls `os.walk` on the local file path. I also assumed more was at play.

Reply:
Before any other decision making: @AlmightyYakob @jwodder do you think it might be valuable to suggest `zarr_checksum` to become **the** implementation of checksumming for zarr within their library?  then we could harmonize/centralize all that to that library instead of maintaining our own code. WDYT?

Reply:
Aside from the bit that this PR fixes, dandi-cli already uses `zarr_checksum` for assembling digests of individual Zarr entries into a checksum for the entire Zarr.  The Zarr-digesting code in dandi-cli itself is largely concerned with iterating over Zarr trees efficiently using threads, including [this thing](https://github.com/dandi/dandi-cli/blob/a51e0a7abb281204e58789facd6ee8914f75beb5/dandi/files/zarr.py#L577), which doesn't really belong in `zarr_checksum`.

Reply:
> Before any other decision making: @AlmightyYakob @jwodder do you think it might be valuable to suggest `zarr_checksum` to become **the** implementation of checksumming for zarr within their library? then we could harmonize/centralize all that to that library instead of maintaining our own code. WDYT?

That would mean that we're once again requiring the `zarr` library, which has heavy dependencies, and whose concern is almost entirely disjoint with our owns (e.g. we use none of the zarr library tools as is for checksumming). So I think it's a bad idea.

Reply:
@yarikoptic 

> do you think it might be valuable to suggest zarr_checksum to become the implementation of checksumming for zarr within their library?

No.  The Zarr checksumming algorithm is something the Dandi project came up with for Dandi's purposes, and I am not aware of the Zarr project expressing any interest in it.

---

What exactly are your remaining concerns, if any, about this PR?

Reply:
> > Before any other decision making: @AlmightyYakob @jwodder do you think it might be valuable to suggest `zarr_checksum` to become **the** implementation of checksumming for zarr within their library? then we could harmonize/centralize all that to that library instead of maintaining our own code. WDYT?
> 
> That would mean that we're once again requiring the `zarr` library, which has heavy dependencies, and whose concern is almost entirely disjoint with our owns (e.g. we use none of the zarr library tools as is for checksumming).

Just for clarity, @jjnesbitt -- didn't you just [above](https://github.com/dandi/dandi-cli/pull/1371/files#r1421063729) confirm using zarr library's construct within zarr_checksum, which is

```
zarr_checksum/generators.py:    from zarr.storage import NestedDirectoryStore
```
which is indeed just a delayed import, but anyways `zarr_checksum` depends on `zarr` anyways and just for this "traverse" functionality?   My point is that ATM `zarr_checksum` does depend on `zarr` and its heavy dependencies, correct?

Reply:
It _currently_ does depend on the `zarr` library, but as we've learned in this discussion, there's no good reason for it. It only depends on the `zarr` library to discover local files to checksum, which could be trivially replaced with `os.walk`.

Reply:
@yarikoptic 

> more efficient threaded walk. Ideally IMHO `zarr_checksum` should just become as efficient, ie carry such an implementation. @jwodder - do you see why that should not or could not be done?

Which efficient threaded walk are you referring to?  [`get_zarr_checksum()`](https://github.com/dandi/dandi-cli/blob/72816ab51a4f168bd9629cc6ac9a6a4edd585bee/dandi/support/digests.py#L101) or the digesting-for-uploading machinery?  I'm fine with the former being in `zarr_checksum`, but the latter currently does not belong there.

> Shouldn't we centralize the walker for zarr, e.g. within zarr_checksum?

Are you going to create an issue about this?  If not, your concern is just going to go into the void.

> So, for now I guess we are doomed to have this duplicated implementation but I think in the longer run we might want to see how we could avoid it.

The entire `ZarrAsset.stat()` code is currently unused by the library.  (I'd look up what commit removed the last use, but I can't get `git log -S` to honor any word boundary syntax I know of.)

Reply:
> The entire `ZarrAsset.stat()` code is currently unused by the library. (I'd look up what commit removed the last use, but I can't get `git log -S` to honor any word boundary syntax I know of.)

something like 47c524c60e4987d36719842645305ecbdaee4627  which seemed was a big refactoring which stopped using `.stat().`?

Reply:
No, that commit only got rid of stat'ing Zarr entries on the Archive.

Reply:
> > more efficient threaded walk. Ideally IMHO `zarr_checksum` should just become as efficient, ie carry such an implementation. @jwodder - do you see why that should not or could not be done?
> 
> Which efficient threaded walk are you referring to? [`get_zarr_checksum()`](https://github.com/dandi/dandi-cli/blob/72816ab51a4f168bd9629cc6ac9a6a4edd585bee/dandi/support/digests.py#L101) or the digesting-for-uploading machinery? I'm fine with the former being in `zarr_checksum`, but the latter currently does not belong there.

We are on the same page here. filed https://github.com/dandi/zarr_checksum/issues/52 

> > Shouldn't we centralize the walker for zarr, e.g. within zarr_checksum?
> 
> Are you going to create an issue about this? If not, your concern is just going to go into the void.

- https://github.com/dandi/dandi-cli/issues/1387